### ISSUE TITLE
Characteristic setValue not reporting back to iOS device

### ISSUE THREAD

**Issue by DerEisvogel** [nfarina/homebridge#493]
Hey everyone,

I have been working with homebridge for more than a month now. Great work! So far everything is working fine for me, apart from problems arising due to my custom plugins for my home.

Experimenting with different services, characteristics and various different approaches on how to map some mechanisms of my home to HomeKit I have come across one minor issue.

Sometimes I wish to just implement a simple Push Button, for example a mute button. Pressing the button once enables mute, pressing it again disables mute. First I tried to use a Stateless Programmable Switches, but those cannot be set from iOS but are rather meant to be set by the devices themselves. I need this to work the other way round. I am now using a switch. Thing is, that there is no On or Off with the mute function, therefore I'd like it to be just a push button. When I turn on the switch the mute function is triggered and a few milliseconds later I reset the switch from within homebridge.

Now, when listening to the 'set' event for 'Characteristic.On' it is required to call the callback function for the value to be actually set. The documentation states, that the value can be changed directly which is working. The 'change' event is triggered, when I change the value to something other the old or the newly set value, and the value is being reported back to HomeKit. But in my case the initial value of the switch is 'false'. When I turn on the switch from iOS it is set to 'true'. In the 'set' listener I would like to be able to call the callback function with 'false'. But when doing that the new value is not being reported back to HomeKit, because the callback only compares the value to the old value.

Wouldn't it be wiser to compare the value from the listener to the newly set value rather than the old cached value?

Basically change the following lines from (Characteristic.js, ll. 168-169):
`        // emit a change event if necessary
        if (oldValue !== actualNewValue)
          this.emit('change', { oldValue:oldValue, newValue:actualNewValue, context:context });`
to:
`        // emit a change event if necessary
        if (newValue !== actualNewValue)
          this.emit('change', { oldValue:oldValue, newValue:actualNewValue, context:context });`

Or maybe another event is needed here, as the value is not really being changed, but needs to be reported back to HomeKit.

Best Regards,
Der Eisvogel

**Comment by KhaosT** [comment]
I believe the ``set`` event callback now only expects an error (and no new value) https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Characteristic.js#L32

**Comment by DerEisvogel** [comment]
Hey KhaosT,

thanks for the quick response. I see, I haven't been up to date then. So the proper way of doing it, is to accept the new value and set a timeout to reset the switch, right?

**Comment by KhaosT** [comment]
@DerEisvogel That should be sufficient for you. Or in other case, I believe you can create a custom characteristic with bool type and write-only permission. Doing so, most HomeKit app will display a stateless touch button for you.

**Comment by DerEisvogel** [comment]
Sounds like an excellent idea, but the apps I am using are not that clever I guess. I have defined my service as follows:
`		Characteristic.PushButton = function() {
			Characteristic.call(this, 'On', '10000025-0000-1000-8000-0026BB765291');
			this.setProps({
				format: Characteristic.Formats.BOOL,
				perms: [Characteristic.Perms.WRITE]
			});
			this.value = this.getDefaultValue();
		};
		inherits(Characteristic.PushButton, Characteristic);


		Service.PushButton = function(displayName, subtype) {
			Service.call(this, displayName, '10000049-0000-1000-8000-0026BB765291', subtype);

			// Required Characteristics
			this.addCharacteristic(Characteristic.PushButton);

			// Optional Characteristics
			this.addOptionalCharacteristic(Characteristic.Name);
		};
		inherits(Service.PushButton, Service);`

I've tried five apps. Elgato Eve was the only one, which at least displayed a basic on off switch, but not a push button, so the state was not being reset. Insteon+, ConnectSense, iDevices and HMCatalog just displayed the device with no options. Did I go wrong anywhere, or have just got the wrong apps? What app are you using?

**Comment by KhaosT** [comment]
@DerEisvogel if you use HMCatalog, you can replace the On characteristic with Identify characteristic (yeah, that's a special case). I'm not sure about other apps but I believe the beta version of "Home" app supports that.

**Comment by DerEisvogel** [comment]
Mh, that doesn't seem to work either. I does result in the right pattern, yes, but that way I cannot include the service in scenes or rooms. I think I'll stick to setting a timeout for now. But anyway, it's interesting how HomeKit can be tricked.

Regarding the 'Home' App: Do you mean this one?
[Home](https://itunes.apple.com/de/app/home-einfache-hausautomatisierung/id995994352?mt=8)
I though about buying that app, but it pretty expensive and I am not sure, whether it's really worth it, Especially because it doesn't support iBeacons like this one for example:
[Kontrol](https://itunes.apple.com/de/app/kontrol-for-homekit/id1051587228?mt=8)

**Comment by KhaosT** [comment]
@DerEisvogel yeah, I guess just stay with timeout then :)
That's the one I'm talking about but if you already have HMCatalog, it's not really necessary to have another similar app.

**Comment by DerEisvogel** [comment]
HMCatalog is not really a nice app, it's just the one that at least supports the features Apple introduced (I think). Isn't Home a little better organized and nicer to work with than HMCatalog?

**Comment by KhaosT** [comment]
It depends on what you want :)
Since you already have a solution, I'll close this issue.

**Comment by DerEisvogel** [comment]
Ok, thanks again for your help!

**Comment by TeraJaQ** [comment]
it's possible have an example of a stateless switch accessory ? i want to made ad irremote for tv...via  MQTT
